skyportal/tests/api/test_annotations.py

```diff
@@ -0,0 +1,296 @@
+        data={
+            'obj_id': public_source.id,
+            'origin': 'kowalski',
+            'data': {'redshift': 0.15},
+            'group_ids': [public_group.id],
+        },
+        token=annotation_token,
+    )
+
+    assert status == 200
+
+    # this should not work, since "origin" Kowalski was already posted to this object
+    # instead, try updating the existing annotation if you have new information!
+    status, data = api(
+        'POST',
+        'annotation',
+        data={
+            'obj_id': public_source.id,
+            'origin': 'kowalski',
+            'data': {'redshift': 0.15},
+            'group_ids': [public_group.id],
+        },
+        token=annotation_token,
+    )
+
+    assert status == 400
+    assert 'duplicate key value violates unique constraint' in data["message"]
+
+
+def test_add_and_retrieve_annotation_group_id(
+    annotation_token, public_source, public_group
+):
+    status, data = api(
+        'POST',
+        'annotation',
+        data={
+            'obj_id': public_source.id,
+            'origin': 'kowalski',
+            'data': {'redshift': 0.15},
+            'group_ids': [public_group.id],
+        },
+        token=annotation_token,
+    )
+    assert status == 200
+    annotation_id = data['data']['annotation_id']
+
+    status, data = api('GET', f'annotation/{annotation_id}', token=annotation_token)
+
+    assert status == 200
+    assert data['data']['data'] == {'redshift': 0.15}
+    assert data['data']['origin'] == 'kowalski'
+
+
+def test_add_and_retrieve_annotation_no_group_id(annotation_token, public_source):
+    status, data = api(
+        'POST',
+        'annotation',
+        data={
+            'obj_id': public_source.id,
+            'origin': 'kowalski',
+            'data': {'redshift': 0.15},
+        },
+        token=annotation_token,
+    )
+    assert status == 200
+    annotation_id = data['data']['annotation_id']
+
+    status, data = api('GET', f'annotation/{annotation_id}', token=annotation_token)
+
+    assert status == 200
+    assert data['data']['data'] == {'redshift': 0.15}
+    assert data['data']['origin'] == 'kowalski'
+
+
+def test_add_and_retrieve_annotation_group_access(
+    annotation_token_two_groups,
+    public_source_two_groups,
+    public_group2,
+    public_group,
+    annotation_token,
+):
+    status, data = api(
+        'POST',
+        'annotation',
+        data={
+            'obj_id': public_source_two_groups.id,
+            'origin': 'kowalski',
+            'data': {'redshift': 0.15},
+            'group_ids': [public_group2.id],
+        },
+        token=annotation_token_two_groups,
+    )
+
+    assert status == 200
+    annotation_id = data['data']['annotation_id']
+
+    # This token belongs to public_group2
+    status, data = api(
+        'GET', f'annotation/{annotation_id}', token=annotation_token_two_groups
+    )
+    assert status == 200
+    assert data['data']['data'] == {'redshift': 0.15}
```

Comment:
```suggestion
    assert data['data']['data'] == {'sdss_redshift': 0.15}
```
Suggest replacing this throughout the test suite with `sdss_redshift` or similar to avoid any confusion about how and where redshifts are set

Reply:
I would prefer to use more realistic examples, both for the data and the origin. Can you think of something more concrete than "kowalski" and "redshift"?

Reply:
Kowalski is fine, and I would use something like `offset_from_host_galaxy`